src/codegen/compile.ml

```diff
@@ -5642,6 +5643,35 @@ module MakeSerialization (Strm : Stream) = struct
       (fun env -> compile_unboxed_const 0l) ^^
     deserialize_from_blob false env ts
 
+(*
+Note [speculating for short (S)LEB encoded bignums]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#3098 highlighted that a lot of heap garbage can be generated while reading in
+(S)LEB-encoded bignums. To make heap consumption optimal for every compactly
+representable `Int`, we resort to speculatively reading a 64-byte chunk from
+the `ReadBuf`. We call it speculative, because it may read past the end of the
+buffer (and thus end up containing junk bytes) or even fail because reading
+across Wasm page boundaries could cause segfaulting. (Consider the buffer ending
+3 bytes before the page boundary and issuing a speculative 64-bit read for the
```

Comment:
```suggestion
3 bytes before the last-memory-page boundary and issuing a speculative 64-bit read for the
```

Reply:
Would it be worth to allocate, in general in the RTS always 8 bytes extra, so that reading a 64bit value can never fail? Would simplify this code (and maybe others), at a very small cost.

Reply:
Your idea (IIRC) is being explored in another PR, where it would come very handy too:  #3090. But I prefer to keep this for now and do the trap suppression there.